server/taskmanager/serializers.py

```diff
@@ -11,13 +21,43 @@ class Meta:
     def to_representation(self, instance):
         """Add dynamic fields"""
         ret = super(PoolSerializer, self).to_representation(instance)
-        if instance.cycle_time is None:
-            ret["cycle_time"] = None
-        else:
+        ret["cycle_time"] = None
+        if instance.cycle_time is not None:
             ret["cycle_time"] = int(instance.cycle_time.total_seconds())
+        ret["max_run_time"] = None
+        if instance.max_run_time is not None:
+            ret["max_run_time"] = int(instance.max_run_time.total_seconds())
         ret["running"] = Task.objects.filter(pool=instance, state="running").count()
         if ret["size"] == 0:
             ret["status"] = "disabled"
+        elif ret["max_run_time"] is not None and ret["cycle_time"] is not None:
+            # Get the run time for the last cycle period and check that it's expected.
+            # New pools will show up as "partial" until a cycle_time has elapsed.
+            duty_cycle = ret["max_run_time"] / ret["cycle_time"]
+            now = datetime.datetime.now(timezone.utc)
+            last_cycle_start = now - instance.cycle_time
+            run_time = 0
+            # We could exclude tasks that are not "success" or "running", but if errors
+            # happen early, they will affect the ratio a lot, and if errors happen late,
+            # fuzzing work probably happened and we should count it.
+            query = (
+                Q(pool=instance) &
+                Q(started__isnull=False) &
+                (Q(resolved__isnull=True) | Q(resolved__gte=last_cycle_start))
+            )
+            for task in Task.objects.filter(query):
+                begin = max(task.started, last_cycle_start)
+                end = task.resolved or now
+                run_time += (end - begin).total_seconds()
+            run_ratio = run_time / ret["cycle_time"]
```

Comment:
Does this need a `float()` somewhere?

Reply:
Yes, it could. However the taskmanager module requires [fuzzing-tc,](https://github.com/MozillaSecurity/fuzzing-tc) which requires Python 3. In Python 3, `float()` would not have any effect. I'll add it for clarity.